drivers/frequency/admv1013/admv1013.c

```diff
@@ -0,0 +1,395 @@
+	buff[1] = data >> 7;
+	buff[2] = data << 1;
+
+	return no_os_spi_write_and_read(dev->spi_desc, buff,
+					ADMV1013_BUFF_SIZE_BYTES);
+}
+
+/**
+ * @brief Reads data from ADMV1013 over SPI.
+ * @param dev - The device structure.
+ * @param reg_addr - The register address.
+ * @param data - Data read from the device.
+ * @return Returns 0 in case of success or negative error code otherwise.
+ */
+int admv1013_spi_read(struct admv1013_dev *dev, uint8_t reg_addr,
+		      uint16_t *data)
+{
+	uint8_t buff[ADMV1013_BUFF_SIZE_BYTES];
+	int ret;
+
+	buff[0] = ADMV1013_SPI_READ_CMD | (reg_addr << 1);
+	buff[1] = 0;
+	buff[2] = 0;
+
+	ret = no_os_spi_write_and_read(dev->spi_desc, buff, ADMV1013_BUFF_SIZE_BYTES);
+	if(ret)
+		return ret;
+
+	/*
+	    |                Byte 0           |     Byte 1     |       Byte 2        |
+	    | 1 | Addr bits 5-0 | Data bit 15 | Data bits 14-7 | Data bits 6 - 0 | 0 |
+	*/
+	*data = ((buff[0] & NO_OS_BIT(0)) << 15 | buff[1] << 7 | buff[2] >> 1);
+
+	return 0;
+}
+
+/**
+ * @brief Update ADMV1013 register.
+ * @param dev - The device structure.
+ * @param reg_addr - The register address.
+ * @param mask - Mask for specific register bits to be updated.
+ * @param data - Data written to the device (requires prior bit shifting).
+ * @return Returns 0 in case of success or negative error code otherwise.
+ */
+int admv1013_spi_update_bits(struct admv1013_dev *dev, uint8_t reg_addr,
+			     uint16_t mask, uint16_t data)
+{
+	uint16_t read_val;
+	int ret;
+
+	ret = admv1013_spi_read(dev, reg_addr, &read_val);
+	if (ret)
+		return ret;
+
+	read_val &= ~mask;
+	read_val |= data;
+
+	return admv1013_spi_write(dev, reg_addr, read_val);
+}
+
+/**
+ * @brief Update Mixer Gate Voltage.
+ * @param dev - The device structure.
+ * @return Returns 0 in case of success or negative error code otherwise.
+ */
+static int admv1013_update_mixer_vgate(struct admv1013_dev *dev)
+{
+	unsigned int mixer_vgate;
+
+	if (dev->vcm_uv < 1800000)
+		mixer_vgate = MIXER_GATE_0_to_1_8_V(dev->vcm_uv);
+	else if (dev->vcm_uv > 1800000 && dev->vcm_uv < 2600000)
+		mixer_vgate = MIXER_GATE_1_8_to_2_6_V(dev->vcm_uv);
+	else
+		return -EINVAL;
+
+	return admv1013_spi_update_bits(dev, ADMV1013_REG_LO_AMP_I,
+					ADMV1013_MIXER_VGATE_MSK,
+					no_os_field_prep(ADMV1013_MIXER_VGATE_MSK, mixer_vgate));
+}
+
+/**
+ * @brief Update Quad Filters.
+ * @param dev - The device structure.
+ * @return Returns 0 in case of success or negative error code otherwise.
+ */
+static int admv1013_update_quad_filters(struct admv1013_dev *dev)
+{
+	unsigned int filt_raw;
+
+	if ((dev->lo_in >= 5400000000) && (dev->lo_in <= 7000000000))
+		filt_raw = LO_BAND_5_4_TO_7_GHZ;
+	else if ((dev->lo_in >= 5400000000) && (dev->lo_in <= 8000000000))
+		filt_raw = LO_BAND_5_4_TO_8_GHZ;
+	else if ((dev->lo_in >= 6600000000) && (dev->lo_in <= 9200000000))
+		filt_raw = LO_BAND_6_6_TO_9_2_GHZ;
+	else
+		filt_raw = LO_BAND_8_62_TO_10_25_GHZ;
```

Comment:
this would execute when dev->lo_in is < 5400000000 or is > 9200000000. dev->lo_in is given in the init param by the user. would it make sense to and another if to check whether dev->lo_in is between 8620000000 and 10250000000 and to add an else which would simply return -1 in case dev->lo_in is not in the expected range?

Reply:
Agree. The out of spec usage of the driver should be independent on the filter selection. I will add a separate condition prior to filter selection.